### DISCUSSION TITLE
Welcome to atol2108 Discussions!

### DISCUSSION THREAD

**Discussion by avgsr** [avgsr/atol2108#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by ShadowSix** [comment]
Привет! Столкнулись с такой же бедой, тысячи чеков за два месяца.

Подскажи пожалуйста, мы этот код в ДТО вставлять должны?

**Reply by avgsr** [reply]
да, ДТО 10.10, или копипаст или кнопочка "открыть"

**Reply by ShadowSix** [reply]
![image](https://github.com/avgsr/atol2108/assets/20943091/3bb91962-8072-4470-9e63-2936d641b92b)
Если кому то будет непонятно. Вот здесь

Хочу выразить вам огромную благодарность за публикацию данного скрипта. В ближ дни эта ветка будет мега популярной, потому что под этот тег папало куча народу((

**Reply by avgsr** [reply]
ещё раз напоминаю, это только идея, код отвратный и под конкретные задачу / данные.
по хорошему нужно переписывать под свои нужды.

**Comment by ShadowSix** [comment]
Еще вопрос: у меня в выгрузке ОФД нету ничего про товарные позиции. Правильно ли я понимаю что в чек коррекции не идут позиции?

Еще в догонку: что именно этот чек корректирует? версию ФФД?

После коррекции чеков нам надо повторно пробивать правильную продажу?

ПС хочется освятить все нуб вопросы, что бы следующим поколениям было легче

**Reply by avgsr** [reply]
> Еще в догонку: что именно этот чек корректирует? версию ФФД?

"чек коррекции" это только название.
что произошло: чеки без тега попали в карантин (нету их) , закрытие смены в налоговой приняли, теперь дизбаланс - движений (чеков) нет а итоги (закрытие смены) есть. налоговая хочет чтоб и движение было.
т.е. чек коррекции то же движение только в доп тегах указывается за какое число и документ это движение.
т.е. коррекция должна быть идентична старому чеку, плюс теги 2108 плюс данные чека который правим (номер дата ФП)

т.е. скрипт делает "чеки коррекции" (движение) сводно по смене, по позициям, с тегами 2108

не знаю как объяснить :(

**Reply by avgsr** [reply]
>После коррекции чеков нам надо повторно пробивать правильную продажу?

нет.
в стандартной ситуации пробивают или когда идут правки, например не тот НДС указали, тогда делают "корректировка возврат прихода" с неправильным НДС плюс "корректировка приход" с правильным, или если не был пробит (вчера чек не пробили, сегодня коррекция приход с указанием вчерашней даты) 
т.е. ситуация с тегом 2108 ближе к "не пробитию", чек коррекции и есть "правильная продажа"

**Reply by Magepal11** [reply]
Жаль, что я увидел вашу историю спустя 2 месяца после её появления. Потому что я бы мог вам ещё в первый день рассказать, что на это не стоит тратить время вообще.
Дам некоторые комментарии:
- "что произошло: чеки без тега попали в карантин (нету их) , закрытие смены в налоговой приняли, теперь дизбаланс - движений (чеков) нет а итоги (закрытие смены) есть. налоговая хочет чтоб и движение было" - бОльшая часть из этого чушь. 
Существует протокол взаимодействия ОФД-ФНС, к которому конечный пользователь кассы никакого отношения не имеет. А вот к чему пользователь имеет отношение, так это к тому, чтобы чеки выгружались в ОФД. Вам не нужно смотреть чеки в ЛК ФНС, только в ЛК ОФД. И ваши чеки там есть. Да, ОФД их отметил, как чеки с ошибками ФЛК (форматно-логического контроля) - ну и что?! Вам до лампочки, что они попали в карантин ФНС - вы сведения о продажах передали, и на этом свои обязанности выполнили;
- согласно упомянутого выше протокола при некритических ошибках ФЛК (некритические - это ошибки, из-за которых ОФД не имеет оснований не прокачивать чек от ККТ) вроде ошибки отсутствия тега 2108 ФНС с помощью ОФД сообщает пользователю, мол, дорогой друг, есть вот такая ошибка, поправь, пожалуйста. И примерно в таком ключе это доносится в большинстве популярных ОФД (кстати, за ofd.ru не скажу, там не видел, как выглядит уведомление об ошибках). Но вот платформа ОФД превзошла по уровню деб...зма всех. Там уведомление читается примерно так "О нет! Твои чеки с ошибками! ФНС их отклоняет! Мы все умрём! Ааа!!!"
- непередача тега 2108 по сути своей является нарушением части 4 статьи 14.5 КоАП - это то же самое нарушение, что и неправильный адрес, неправильная ставка НДС, отклонение времени более чем на 5 минут. Только, извините меня, все "кладут" на эти нарушения, но от непередачи 2108 обкакались. Всё потому, что 2108 - "распиаренное" нарушение. Судя по моему опыту, как работает ФНС с нарушениями, по которым когда-то было сказано "вот вчера вы били чеки так, а сегодня должны вот так", никого за это нарушение штрафовать не будут. Яркий пример: с 1 марта 2019 года всем "строго-настрого" запретили работать с ФФД1.0, рассказывали сказки, как ОФД будут блокировать передачу чеков, как все пользователи, кто не перешёл на ФФД1.05, попадут через 30 дней на внеплановую замену ФН, а дополнительно - на штраф от налоговой. Что на деле? Никого с 1 марта не заблокировали, и в осенью 2020 года к нам обращались пользователи с кассами, всё ещё работающими на ФФД1.0;
- да, я не отрицаю, что была массовая рассылка со стороны ФНС о том, что кассы работают с нарушениям. При этом ни в одном Информационном письме от ФНС не было указано конкретного нарушения - все Информационные письма, что я видел до этого, всегда содержали не только указание, на какой кассе нарушение, но и конкретно что нарушено. Сейчас такого нет. Обращаю внимание, что Информационное письмо это не Предписание об устранении нарушений (недостатков), оно не налагает на вас обязательств по выполнению каких-либо действий, будь то чек коррекции или что-то другое. Дааа, закон предписывает при ошибках с реквизитами выполнять коррекции таких чеков, а именно
1) пробить чек коррекции возврата прихода с некорректным реквизитом;
2) пробить чек коррекции прихода с корректным реквизитом.
Но с учётом предыдущего моего комментария - это просто трата времени. У нас есть пользователь не на договоре с 26 точками. У него по ОФД светились ошибки по 6 кассам. Чеков с ошибками с конца ноября по февраль, когда я обсуждал 2108 с представителем ОФД Такском, - более 15 тысяч. Никаких чеков коррекции пробивать не будем. И я вам гарантирую, что ни штрафов пользователь не получит, ни даже объяснительных в налоговую писать не будет.
Как специалист, набивший руку на выполнении чеков коррекции вручную (на выполнение одного чека я трачу в районе 1 минуты), могу вас уверить, что скрипты вам не помогут сэкономить время, т.к. по каждому чеку, даже если вы будете пробивать 1 за смену (в реальности придётся пробивать 4 за смену (возврат/приход нал/безнал) вам нужно менять сумму и дату/время, а это всё равно придётся делать вручную.
Поэтому мой вам совет: если в нынешних чеках у вас ошибки нет (вы обновили драйвер/прошивку/версию вашего кассового ПО), то все предыдущие корректировать нет смысла. Не тратьте время

**Comment by avgsr** [comment]
> Еще вопрос: у меня в выгрузке ОФД нету ничего про товарные позиции. Правильно ли я понимаю что в чек коррекции не идут позиции?

у нас "первый офд", там есть отчет "чеки с товарными позициями", от него и танцевал, делал одна смена  = один чек, свернуто по позициям, "приход" плюсом, "возврат прихода" минусом.
можно попробовать собрать данные или из учетной программы или из Фискального Накопителя (только я не уверен что они там есть все).

**Reply by ShadowSix** [reply]
Прикрепите пожалуйста пример CSV файла. В котором два-три чека. 

Тоже будет полезно тем кто ищет, потому что ОФД разные

**Reply by avgsr** [reply]
в скриптах:
function file2recs(file) {
//123123; 573; 2020-01-30T21:55; Приход.; 35; 0; Товар; Наим Товара; 35.00; 1
	var col = {"fd":0,"sh":1,"dt":2,"tr":3,"ch":4,"el":5,"p2":6,"nm":7,"pr":8,"qn":9};

объект col - индексы столбцов, нумерация с нуля
fd: номер чека, индекс 0
sh: номер смены, 1
dt: дата чека, 2
tr: тип чека , "Приход." или "Возврат прихода."
ch: оплачено наличными по чеку
el: оплачено безналичными по чеку
p2: тип предмета, "Товар" "Услуга" ...
nm: наименование предмета
pr: цена
qn: количество

если у Вас другой порядок или количество столбцов то нужно поправить индексы на правильные (допустим: номер смены - пятый столбец, вместо "sh":1 поправить "sh":4 )

файл отсортирован по номеру смену и номеру чека

**Reply by ShadowSix** [reply]
Разобрался.

//123123; 573; 2020-01-30T21:55; Приход.; 35; 0; Товар; Наим Товара; 35.00; 1

в упор не видел, потом сделал по аналогии и увидел ОК в логах

**Reply by ShadowSix** [reply]
Вылезла проблема. В чек коррекции вылезла копейка.

9225;350;2023-11-09T14:34;Приход.;0;1844;Товар;Наименование позиции;18.43;7

Как я понял скрипт перемножает цену на колво и суммирует по позициям

Общая стоимость позиции с учетом скидок и наценок = 18.43*7=129.01

Можете помочь добавить округление? Что бы любые копейки округлялись до ровной суммы.

**Reply by avgsr** [reply]
или
https://integration.atol.ru/api/#zaregistrirovat-pozitsiyu
LIBFPTR_PARAM_POSITION_SUM - полная сумма позиции. Может отличаться от произведения цены на количество. В этом случае ККТ разобьет позицию на две, распределив получившуюся разницу (скидку или надбавку) между позициями.

или

Регистрация итога чека
Для регистрации итога чека требуется вызвать метод receiptTotal(), передав сумму чека в параметре LIBFPTR_PARAM_SUM.
Допускается регистрация итога меньше суммы чека, но только в рамках копеек. Т.е., если общая сумма чека равна 1000.57, то можно зарегистрировать итог от 1000.00 до 1000.57.
Метод не является обязательным. Если его не использовать, сумма чека будет посчитана автоматически, без округлений копеек.

Скорее всего Вас больше интересует все-таки итог по чеку без копеек чем итог по позиции?

как вариант: сделать отдельный столбик в cvs с итогом по чеку: 
//9225;350;2023-11-09T14:34;Приход.;0;1844;Товар;Наименование позиции;18.43;7; 1844
	var col = {"fd":0,"sh":1,"dt":2,"tr":3,"ch":4,"el":5,"p2":6,"nm":7,"pr":8,"qn":9 , "tt":10}; //tt -итог по чеку
...
			curRec.elec = 0;
			curRec.total = 0;
...
				curRec.elec += new Number(arStr[col.el])*k;
				curRec.total += new Number(arStr[col.tt])*k; //итог чека 
...
Fptr.setParam(Fptr.LIBFPTR_PARAM_SUM, 1.00*rec.total);
Fptr.receiptTotal();
res = Fptr.closeReceipt(); /* выполняем метод closeReceipt() для закрытия чека */

или считать как проверяется chek_cvs  (если сумма оплат = сумма позиций то ок), итог по чеку поставить сумму оплат
			curRec.total = 0;
...
				curRec.total += new Number(arStr[col.ch])+new Number(arStr[col.el]); //итог чека сумма оплат
...
Fptr.setParam(Fptr.LIBFPTR_PARAM_SUM, 1.00*rec.total);
Fptr.receiptTotal();
res = Fptr.closeReceipt(); /* выполняем метод closeReceipt() для закрытия чека */

Если округлить по позиции:
не знаю как округлить в Duktape. Погуглите "javascript округление", может какая функция и найдется...
но можно целочисленное деление  % (отбросит дробную часть):
					com.pr = new Number(arStr[col.pr]);
					com.qn = k*new Number(arStr[col.qn]);
					com.tt = (com.pr*com.qn)%1;
...
		Fptr.setParam(Fptr.LIBFPTR_PARAM_QUANTITY, 1.00*com.qn);
		Fptr.setParam(Fptr.LIBFPTR_PARAM_POSITION_SUM, 1.00*com.tt); 

как то так, пробовать не на чем, подозреваю что все варианты не правильные, пробуйте сами, удачи :)

**Comment by ghost** [comment]
Добрый день. Подскажите, пожалуйста, сейчас попробовали 10 строк csv прогрузить. Получился 1 чек с 10 позициями. А можно сделать так, что бы это был на 1 чек, а 10 разных чеков? То есть одна строка в CSV - это один отдельный чек

**Reply by avgsr** [reply]
ничего не понял :(
Постановка задачи была 1 чек на смену, у Вас видимо другая постановка. Образец csv хотя бы чтоб понятнее было что хотите.

**Reply by ghost** [reply]
у нас постановка делать 1 чек с одним товаром из csv. CSV привели к такому же виду как у вас. То есть скрип должен брать 1 строку, делать отдельный чек и отправлять его, далее следующий по списку из csv и тд. Но уже не актуально, получилось на основе вашего скрипта сделать, теперь работает как надо) Спасибо
